### DISCUSSION TITLE
Tab Autocomplete (line-autocomplete, such as Cursor Tab)

### DISCUSSION THREAD

**Discussion by arturskruze** [cline/cline#1083] [Feature Requests]
Cline is superior when it comes to task editing but there isn't a good alternative for Cursor Tab.

**Comment by MentalGear** [comment]
I was also surprised by the lack of this often more used feature than full code generation, and wondering how difficult it is to add this?
A naive first version would only need the context of the file/previous text before the cursor.

**Comment by quezadaesteban** [comment]
Maybe they could get some inspiration from Conitnue's VSCode extension, which supports tab autocomplete through local or third-party models.

**Reply by farmerinatechstack** [reply]
Are you using that extension? How's it going?

**Comment by alioshr** [comment]
Imo this is pretty hard to accomplish with Cline's concept. Cline is a middleman, using models. A good autocomplete like cursor demands custom model training, etc. 

You cannot get a custom model like this from openrouter. 

It is not viable to use local models for this, at least for the majority of the ppl. 

So you would need to host a model like this. Then this goes totally in a different direction from the idea here.

It would be amazing if we had a marvelous framework/model for an astonishing autocomplete hosted somewhere and it would cost something < 10usd per month & be open source w/ the possibility to self host. 

Then I would say bye bye to cursor.

**Reply by farmerinatechstack** [reply]
Are you saying that the Cursor build goes deeper, whereas Cline is just sort of a "middleman" between us and the IDE? Is there anywhere you'd recommend I can check out to learn more? I'm just diving into this since I'm revisiting my dev setup.

**Reply by Ken-Andre** [reply]
I'm thinking that more vibecoders than traditionnal coders(i mean those who studied programming and start from there before using ai to perform) use cline. This only can explain why most still don't understand why for now we cannot have autocomplete here. It's my hypothesis btw...

**Reply by KeroZhai** [reply]
> I'm thinking that more vibecoders than traditionnal coders(i mean those who studied programming and start from there before using ai to perform) use cline. This only can explain why most still don't understand why for now we cannot have autocomplete here. It's my hypothesis btw...

That's the future, but for now I think traditional coders are stiil the majority, at least in enterprises.

**Reply by Panossa** [reply]
I agree with @KeroZhai here, but that's beside the point. I don't think you really need a specific model for autocomplete.

There's two routes already possible without it: 
1. Via the old FIM-style requests with prefix and suffix (which OpenAI used up until a 4.X version and Deepseek and some others still use)
2. Literally just sending a "full" request to whatever model is connected with the surrounding context. The exact context would need to be tweaked over time, but just sending the surrounding method(?) or X lines would already be better than nothing.

Although I must say I don't know if the Cline team is really doing anything besides developing the new (barely needed) desktop app, so I guess we'll have to wait until we get relevant updates. They might need more people working on the project if they're developing a CLI, a plugin, a desktop app, and an SDK at the same time, or else they'll be left in the dust.

**Comment by radosek** [comment]
@alioshr this is why you can hook it up to mistral codestral api (or locally with ollama) in Continue. It would work fine in the same way in Cline.

**Reply by alioshr** [reply]
Hi! In my personal experience these solutions are miles away from Cursor's autocomplete.

**Reply by redelman431** [reply]
Why does it need to be like Cursor's? What's wrong with using Continue's more BYOB third party API solution to autocomplete? It may not be as robust as Cursor's solution but I think it is good enough. Why not just settle with that for now?

**Comment by MentalGear** [comment]
Agreed that tab autocomplete should be prioritised to get feature parity with tools like Windsurf/Cursor that have the only differentiator left being tab auto complete.

**Reply by abrahem79** [reply]
[https://app.luvrio.app/tasktracker/tasks/442e246f-779d-49b8-b1f8-b2f07fd9508e]()

**Comment by hasban12138** [comment]
+1

**Comment by jakobsturm** [comment]
I think this will become much easier to do now that VS Code opens up some parts of its Co-Pilot extension

**Comment by buildgreatthings** [comment]
+1

**Comment by gfylol** [comment]
+1

**Comment by Lijbers** [comment]
+1

**Comment by mborn1** [comment]
+1

**Comment by MentalGear** [comment]
👍 On 13 Jul 2025, at 17:44, Mathias Born ***@***.***> wrote:﻿
+1

**Comment by jan-hlavsa-mindplexic** [comment]
+1

**Comment by yamirghofran** [comment]
+1

**Comment by Cwagne17** [comment]
+1

**Comment by chen-ziwen** [comment]
Augment is also a plug-in, but it provides a complete function.

**Comment by alioshr** [comment]
is a neovim plugin to be taken under consideration if such feature is created? Cline is OSS and Neovim as well. It would be nice.

**Comment by chen-ziwen** [comment]
OK，thanks



---Original---
From: "Aliosh ***@***.***&gt;
Date: Mon, Aug 18, 2025 20:43 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [cline/cline] Tab Autocomplete (line-autocomplete, such as CursorTab) (Discussion #1083)




 
is a neovim plugin to be taken under consideration if such feature is created? Cline is OSS and Neovim as well. It would be nice.

**Comment by yohplala** [comment]
+1

**Comment by luobote55** [comment]
+1

**Comment by M0ritzFeickert** [comment]
+1

**Comment by MarkLee425** [comment]
+1

**Comment by josh900** [comment]
+1

**Comment by RaymondDunn** [comment]
+1

**Comment by rsoorajs** [comment]
+1

**Comment by epicwhale** [comment]
+1

**Comment by david-AH** [comment]
+1

**Comment by lesca** [comment]
+1

**Comment by pmarxbraun** [comment]
+1

**Comment by eret9616** [comment]
+1

**Comment by smilelikeshit** [comment]
+1

**Comment by Ldes** [comment]
Not best, but works - line-autocomplete with https://supermaven.com/

**Comment by shlomiLan** [comment]
+1

**Comment by lonnie776** [comment]
+1

**Comment by brat002** [comment]
+1

**Comment by shloimy-wiesel** [comment]
+1

**Comment by grandones** [comment]
+1

**Comment by KeroZhai** [comment]
+1 for autocomplete

**Comment by ACheshirov** [comment]
+1 for autocomplete. Continue.dev has it, so why can't it be added here too? Cline simply works better on the agent mode but I'm using Continue.dev because of the autocomplete.

**Reply by KeroZhai** [reply]
Continue.dev does has autocomplete, but for me it doesn't works very well. And sadly, there seems no alternatives.

https://github.com/continuedev/continue/issues/8661

**Reply by Panossa** [reply]
@KeroZhai an alternative is GitHub Copilot, but it only works with (at best) GPT 4.1 or something similar, so it's nowhere close to Cursor's performance, sadly. It's also $10/month+.

**Reply by redelman431** [reply]
Continue's autocomplete performance depends on what models you attach to it and how you configure and tune it.

**Comment by realyasster** [comment]
Autocomplete seriously important feature. Cline must add that.

**Comment by lgmadurange** [comment]
+1

**Comment by MarkMuravev** [comment]
+1

**Comment by sPredictorX1708** [comment]
+1

**Comment by abrahem79** [comment]
[https://app.luvrio.app/tasktracker/tasks/442e246f-779d-49b8-b1f8-b2f07fd9508e]()

**Comment by Ken-Andre** [comment]
Really?
I thinked that enterprise used more Claude code or Kiro.


KENMOGNE ANDRE

**Comment by fomnn** [comment]
+1

**Comment by PoulLorca** [comment]
+1

**Comment by B33fb0n3** [comment]
+1

**Comment by khimaros** [comment]
i created this as a minimal auto complete extension for all OpenAI compatible endpoints. also supports llama.cpp FIM syntax: https://github.com/khimaros/mortar

**Comment by TimothyFran** [comment]
+1

**Comment by luandnh** [comment]
+1

**Comment by redelman431** [comment]
Considering that the current version of Cline utilizes a one size fit all configuration solution for act and plan mode functions and has the hard requirement of chain of thought models in both modes. I think adopting a more granular system of model selection/configuration similar to the Continue extension would be an important pre-requisite to paving the way for features such as autocomplete. Especially considering that autocomplete models have technical requirements opposite to Cline's general AI paradigm, where they must NOT have chain of thought, take simple prompts instead of rigid system prompts, and act immediately without plan. So I believe the path of least resistance for such a feature would be to work on this feature request first! 

See
https://github.com/cline/cline/discussions/2465


Basically add similar model selection/configuration granularity that resembles the  Continue IDE extension. That way we can have the flexibility to configure technically suitable models for more specific AI functions.  Except unlike Continue,  have guards in the system that require COT models for functions where it is essential to follow Cline's strict COT autonomous agentic design paradigm, but have more lenient guards for specific functions where the autonomous agent COT paradigm isn't essential or counterproductive. This would make the design of autocomplete and many other features more simple down the road by removing a lot of tight coupling. Feel free to agree or disagree.

**Reply by Panossa** [reply]
I mean, we already have the Plan and Act tabs where we can set a model and define the thinking mode. Similarly, you could add new tabs like "autocomplete" and remove the thinking dropdown for it. Doesn't seem like it needs that much thought, tbh,

**Reply by redelman431** [reply]
I guess perhaps you are right, we may not need the extra step of model selection granularity to resemble Continue. We could  just simply make a new specific configuration exclusively for autocomplete that doesn't have the same paradigm guards that Plan and Act mode have. Perhaps adding Continue style model selection granularity all the way down to editing, transformers, and embeddings too early may make testing more complicated and allow more places for this application to break. So that is indeed a valid point. I just wasn't entirely sure how coupled the paradigm guards are to the entire app, whether it was coupled to the point of redesigning a lot of the app or if we can simply extend the autocomplete model selection widget to not have the same model selection paradigm guards as the plan and act model select widgets. Just have the autocomplete model configuration dropdown do the opposite of what the plan and act configuration do and limit only to AI models with very high performance and no COT that are compatible to predictive code typing.  But still the rather similar Continue extension is good conceptual proof that autocomplete is technically feasible in Cline. It could also be a good inspiration on how we can effectively implement the autocomplete feature quickly and efficiently using a similar configurable third party BYOB architecture.

**Comment by khimaros** [comment]
i built https://github.com/khimaros/mortar as a simple (FIM) autocomplete for llama-server

works with anything that offers the OpenAI Completions endpoint
